G1-true-binary-metadata.md

```diff
@@ -0,0 +1,89 @@
+True Binary Metadata
+----
+* Author(s): ctiller
+* Approver: ejona
+* Status: Draft
+* Implemented in: n/a
+* Last updated: 3/29/17
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Ww8waYz_Nes
+
+## Abstract
+
+Add a HTTP2 extension to allow binary metadata (those suffixed with -bin) to be
+sent without the base64 encode/decode step.
+
+## Background
+
+gRPC allows binary metadata to be sent by applications. These metadata elements
+are keyed with a -bin suffix. When transmitted on the wire, since HTTP2 does not
+allow binary headers, we base64 encode these elements and Huffman compress them.
+On receipt, we transparently reverse the transformation.
+
+This transformation is costly in terms of CPU, and there exist use cases where
+this transformation can become the CPU bottleneck for gRPC.
+
+### Related Proposals:
+
+n/a
+
+## Proposal
+
+### New setting
+
+Expose a custom setting in our HTTP2 settings exchange:
+GRPC_ALLOW_TRUE_BINARY_METADATA = 0xfe03.
```

Comment:
So gRPC uses -bin to specify that a header is binary, and says that other metadata must be ascii (per http2). That -bin has semantic meaning (foo-bin is a different header than foo).

We use that as a signal to base64 encode things right now. Under this proposal we'd just switch to an alternate encoding.

Since we can't universally make all of the HTTP2 ecosystem take binary metadata, a mechanism to allow transmitting anything and dropping the -bin isn't possible for us (we need to keep interoperating - so proxies that support this would potentially need to do the base64 conversion if one side didn't support this mechanism).

If there were an extension that allowed sending binary values in metadata, we'd still end up needing a disambiguating byte for backwards compatibility (for the gRPC -bin suffixed headers), and we'd still maintain the requirement that non -bin headers must be ascii.